### PR TITLE
ci(publish): add initial publish app GH workflow

### PR DESCRIPTION
**Pull request by h1dden-da3m0n**
### Description

This is an initial attempt on offering a replacement for the sol remaining Azure CI workflow.
Any feedback is welcome and the are still at least 2 steps missing from the original workflow (namely: `CopyFilesOverSSH`, `SSH`, mostly since I don't know what they are exactly used for or how to replicate them).

### Changes

* Add publish GitHub workflow 

### Issues

* n/a (move from Azure CI to GH Actions)

### CODE REVIEW

File: .github/workflows/publish.yaml
Comment (nielsvanvelzen) [inline .github/workflows/publish.yaml:31]: Should just set the env var so all gradle tasks pick it up automatically, like so:

https://github.com/jellyfin/jellyfin-sdk-kotlin/blob/5d1d7e3600d8f9d2f98dd5dbc2a68f4090f4f416/.github/workflows/sdk-publish.yaml#L22-L25

File: .github/workflows/publish.yaml
Comment (nielsvanvelzen) [inline .github/workflows/publish.yaml:1]:
```suggestion
name: App Publish
```

Also, file name should be "app-publish.yaml"

File: .github/workflows/publish.yaml
Comment (nielsvanvelzen) [inline .github/workflows/publish.yaml:5]:
```suggestion
    tags:
      - v*
```

**Review by nielsvanvelzen — CHANGES_REQUESTED** [review]
(Only reviewed until the signing part)

Ideally the publishing workflow also publishes test versions (on the master branch) that are not signed. The Kotlin SDK works similarly. See https://github.com/jellyfin/jellyfin-sdk-kotlin/blob/5d1d7e3600d8f9d2f98dd5dbc2a68f4090f4f416/.github/workflows/sdk-publish.yaml

File: .github/workflows/publish.yaml
Comment (h1dden-da3m0n) [reply .github/workflows/publish.yaml:31]: Thy for the reference and given the different application than I assumed I agree that the env var will work better.
I used this method originally to make the use of the version more transparent (like we discussed in the test PR ages ago :sweat_smile:).

Comment (h1dden-da3m0n) [reply .github/workflows/publish.yaml:31]: Allow me to ask this though: Why do you need the `tr / -` pipe in the sdk repo?

Comment (nielsvanvelzen) [reply .github/workflows/publish.yaml:31]: The `tr` part is not necessary but I kept it just to be safe IIRC

Comment (h1dden-da3m0n) [reply .github/workflows/publish.yaml:31]: mind if I omit it? additionally should I too remove the `v` prefix or let the VersionUtils.kt handle it?

Comment (nielsvanvelzen) [reply .github/workflows/publish.yaml]: I'd rather keep it in case there was a reason that I forgot :p

Also the v prefix can be used to detect version tags in the CI but it should **always** pass the exact tag to the Gradle build scripts. The Gradle build scripts will handle it.

Comment (h1dden-da3m0n) [reply .github/workflows/publish.yaml]: Okay then I'll use your version of the export 1:1. Regarding the tag I agree, however was unsure since the old script handled only the `refs/tags` bit so wanted to be sure :+1:

**Comment by h1dden-da3m0n** [comment]
okay, I am not 100% on the repo upload script NGL. Please let me know what you think of that, otherwise this *should* cover the requirements to this workflow.
There are a few secrets that would be required to be added to the repo should the workflow in its current form be accepted, but I expect this to still go throug an iteration or two.

Note I plan to squash the workflow commits in the end unless you want the full history, but this should be discussed at the end.

**Comment by h1dden-da3m0n** [comment]
okay I opted to remove my shell script in favor of two actions, I found them to be more straight forward (however I can always revert if you like the script more)
There are several secrets that this workflow would require to be added to this repo before it would work and I still need to squash the workflow commits, but functionality wise it should be equivalent to the Azure CI pipeline.

File: .github/workflows/app-publish.yaml
Comment (nielsvanvelzen) [inline .github/workflows/app-publish.yaml:9]:
```suggestion
      - master
```

No need to run on release branches because we (normally) only push when we're going to tag it.

File: .github/workflows/app-publish.yaml
Comment (nielsvanvelzen) [inline .github/workflows/app-publish.yaml:32]: Don't use caching for the publish task, always build clean so the cache can't cause or suppress any issues.

File: .github/workflows/app-publish.yaml
Comment (nielsvanvelzen) [inline .github/workflows/app-publish.yaml:39]:
```suggestion
      - name: Build Release
        run: ./gradlew --no-daemon --info assembleRelease versionTxt
```

File: .github/workflows/app-publish.yaml
Comment (nielsvanvelzen) [inline .github/workflows/app-publish.yaml:43]:
```suggestion
          mkdir build/jellyfin-publish/release-libre build/jellyfin-publish/release-proprietary;
```

I'd rather use a (custom) build directory in the root

Comment (nielsvanvelzen) [reply .github/workflows/app-publish.yaml:43]: The version.txt should be moved to `build/jellyfin-publish/`, and that directory will be the artifact path

File: .github/workflows/app-publish.yaml
Comment (nielsvanvelzen) [inline .github/workflows/app-publish.yaml:59]:
```suggestion
    needs: build
```
The job is named `build`

File: .github/workflows/app-publish.yaml
Comment (nielsvanvelzen) [inline .github/workflows/app-publish.yaml:88]: We might want to move the signing part to Gradle, We'd need to apply the `signingConfigs` in the `android {}` block for that though.

File: .github/workflows/app-publish.yaml
Comment (nielsvanvelzen) [inline .github/workflows/app-publish.yaml:91]: Can you change it to use `alexellis/upload-assets@0.2.2`? Helps with consistency between the various Android/Kotlin projects (Kotlin SDK uses that one)

File: .github/workflows/app-publish.yaml
Comment (nielsvanvelzen) [inline .github/workflows/app-publish.yaml:60]:
```suggestion
    environment: release
    if: ${{ startsWith(github.ref, 'refs/tags/v') }}
```

The secrets will be added to the release environment

**Review by nielsvanvelzen — CHANGES_REQUESTED** [review]

File: .github/workflows/app-publish.yaml
Comment (h1dden-da3m0n) [reply .github/workflows/app-publish.yaml:91]: NGL, with all the various versions of Actions you can find to do one thing this srsly gets rather convoluted and difficult to track which version is 1. maintained and 2. security patched.
The reason why I went with this one was due to its wider feature set (and since the official GH `actions/upload-release-asset` was deprecated :disappointed: )

However, I agree if you use the other action in the Kotlin SDK repo I will switch to that one :wink:

File: .github/workflows/app-publish.yaml
Comment (h1dden-da3m0n) [reply .github/workflows/app-publish.yaml:88]: :thinking: I understand what you mean, however have to disclose I would have no idea on how to do that. My experience with Gradle so far is rather limited, but I will see what I can do :+1:

Comment (nielsvanvelzen) [reply .github/workflows/app-publish.yaml:88]: No problem, we can always change this part later

File: .github/workflows/app-publish.yaml
Comment (h1dden-da3m0n) [reply .github/workflows/app-publish.yaml:32]: Sounds reasonable, I have not thought about it this way NGL

File: .github/workflows/app-publish.yaml
Comment (h1dden-da3m0n) [reply .github/workflows/app-publish.yaml]: Okay I read up on this a bit and have a couple of questions:

* given the logic of this, I assume the jks for the gradle based solution still needs to be base64-ed and serialized back to a jks before the build?
* given what I have found the signing would happen as part of the release build, if the jks, its passwd and the alias where present and unsigned if they are missing. However, I assume you thought of a custom gradle signing task that does nothing but sign the output?

Comment (Maxr1998) [reply .github/workflows/app-publish.yaml]: That's correct, you'd deserialize the keystore, put it to the right location, set the alias and passwords from the secrets, and the signing is done by the build tools. I did exactly that a while ago, see [here](https://github.com/Maxr1998/MaxLock/blob/master/buildSrc/src/main/kotlin/SigningConfigHelper.kt#L24) for reference.

Comment (h1dden-da3m0n) [reply .github/workflows/app-publish.yaml]: thank you for that reference, I shall give it a try to integrate it here.

Comment (h1dden-da3m0n) [reply .github/workflows/app-publish.yaml]: Okay, 412a4d58414d5f70f58206b9bcd679f1964a0416 now has a very 'scripted' variation of the sign config contained in the build.gradle.kts. However, I guess you imagined something more elegant than that, so please let me know.

Comment (h1dden-da3m0n) [reply .github/workflows/app-publish.yaml]: Okay, ended up installing Android Studio because IDEA refused to import properly and added a very 'scripted' signing config in a95b2e4c75adf799615286cfec5941dbacb563e4. I assume you where thinking of a more elegant solution, so let me know if I should change something or just remove the commit an squash the rest of the workflow commits.

File: .github/workflows/app-build.yml
Comment (nielsvanvelzen) [inline .github/workflows/app-build.yml:40]:
```suggestion
          name: build-artifacts
          retention-days: 14
```

File: .github/workflows/app-publish.yaml
Comment (nielsvanvelzen) [inline .github/workflows/app-publish.yaml:31]: - [ ] These need to be added (preferably on organization level so we can re-use it in the ATV app later).

_(mark when added)_

File: .github/workflows/app-publish.yaml
Comment (nielsvanvelzen) [inline .github/workflows/app-publish.yaml:59]: - [ ] These need to be added (preferably on organization level so we can re-use it in the ATV app later).

_(mark when added)_

**Review by nielsvanvelzen — CHANGES_REQUESTED** [review]

File: .github/workflows/app-build.yml
Comment (h1dden-da3m0n) [reply .github/workflows/app-build.yml:6]: I mean the test workflow already runs on PRs IIRC,  however I see the appeal of having a per-compiled artefact to test b4 merging a pr

File: .github/workflows/app-build.yml
Comment (h1dden-da3m0n) [reply .github/workflows/app-build.yml:25]: I can remove this again, no problem, but I will wait for @Maxr1998's comment regarding it since (IIRC) he liked the idea

Comment (Maxr1998) [reply .github/workflows/app-build.yml:25]: I don't have an opinion on this, so if Niels wants it removed, we should do that ^^

Comment (h1dden-da3m0n) [reply .github/workflows/app-build.yml]: gotcha and it went pooof :ghost:

File: .github/workflows/app-build.yml
Comment (nielsvanvelzen) [inline .github/workflows/app-build.yml:29]: We need the debug versions here or just all .apk files. An unsigned release can't be installed on a device.

Comment (h1dden-da3m0n) [reply .github/workflows/app-build.yml:29]: would that not require a "debug-keystore" to be passed into the build step of the `app-build` workflow?

**Review by nielsvanvelzen — APPROVED** [review]
Note: do not merge until secrets are added.

File: .github/workflows/app-publish.yaml
Comment (anthonylavado) [reply .github/workflows/app-publish.yaml:31]: What are KEY_ALIAS and KEY_PASSWORD? I have added KEYSTORE and KEYSTORE_PASSWORD for this repo and Android TV.

Comment (anthonylavado) [reply .github/workflows/app-publish.yaml:31]: After further review, KEY_ALIAS has been added. I believe KEY_PASSWORD is unused here.

File: .github/workflows/app-publish.yaml
Comment (nielsvanvelzen) [inline .github/workflows/app-publish.yaml:11]: environment is not used now, sorry :p

**Review by nielsvanvelzen — CHANGES_REQUESTED** [review]

**Review by nielsvanvelzen — APPROVED** [review]

**Review by Maxr1998 — APPROVED** [review]
LGTM. Waiting for all secrets to be added before merging.

**Review by nielsvanvelzen — APPROVED** [review]

**Pull request merged**
